### PR TITLE
[SimplifyLibCall][Attribute] Fix bug where we may keep `range` attr with incompatible type

### LINKED ISSUE #112633: [InstCombine] Crash on valid input when narrowing type but not updating `range` attribute
https://godbolt.org/z/b3M1dj6aj

```
define void @SetRow_C(ptr %p, i32 %conv) {
entry:
  %call.i = tail call ptr @__memset_chk(ptr %p, i32 range(i32 0, 123) %conv, i64 1, i64 1)
  ret void
}

declare ptr @__memset_chk(ptr, i32, i64, i64)
```

We keep the `range(i32 ...)` when we replace the `_chk` with `memset.inline.p0.i64` which leads to a crash:
```
Range bit width must match type bit width!
  tail call void @llvm.memset.p0.i64(ptr noundef nonnull align 1 dereferenceable(1) %p, i8 range(i32 0, 250) %0, i64 1, i1 false)
```

### PR DESCRIPTION
**Pull request by goldsteinn**
In a variety of places we change the bitwidth of a parameter but don't
update the attributes.

The issue in this case is from the `range` attribute when inlining
`__memset_chk`. `optimizeMemSetChk` will replace an `i32` with an
`i8`, and if the `i32` had a `range` attr assosiated it will cause an
error.

Fixes #112633

### CODE REVIEW

File: llvm/include/llvm/IR/Attributes.h [hunk 1]
```diff
@@ -1292,7 +1292,8 @@ bool isNoFPClassCompatibleType(Type *Ty);
 /// if only attributes that are known to be safely droppable are contained in
 /// the mask; only attributes that might be unsafe to drop (e.g., ABI-related
 /// attributes) are in the mask; or both.
```
Comment (nikic) [inline llvm/include/llvm/IR/Attributes.h:1294]: Update doc comment for AS parameter. In particular, it should be clear that the returned mask is not required to be a subset of AS. AS is effectively just a type hint.

File: llvm/include/llvm/IR/InstrTypes.h [hunk 2]
```diff
@@ -1453,14 +1453,26 @@ class CallBase : public Instruction {
   /// looking through to the attributes on the called function when necessary).
   ///@{
 
-  /// Return the parameter attributes for this call.
+  /// Return the attributes for this call.
   ///
```
Comment (nikic) [inline llvm/include/llvm/IR/InstrTypes.h:1457]: While you're already here, please also drop these empty `///`.

**Review by nikic — APPROVED** [review]
LGTM

This is quite ugly, but I don't have a better suggestion on how to handle this generically.

**Comment by goldsteinn** [comment]
> LGTM
> 
> This is quite ugly, but I don't have a better suggestion on how to handle this generically.

Agreed. I think there are some helpers you could wrap this with to clean up some the uses, but AFAICT this was necessary.

**Comment by goldsteinn** [comment]
@nikic, I verified that this is an issue on 19. Think for the backport I should just handle the case in SimplifyLibcall, or backport the whole thing? My guess is that there are other manifestations of this bug (outside of SimplifyLibcall), but its not a particularly clean thing to backport.

**Comment by nikic** [comment]
Without the inlining change, I don't think there's a pressing need to backport this.

**Comment by goldsteinn** [comment]
> Without the inlining change, I don't think there's a pressing need to backport this.

Okay, although the test in this PR (without inlining) does crash opt on 19 release.

**Pull request merged**